### ISSUE TITLE
Think about nested quasiquotes

### ISSUE THREAD

**Issue by Technologicat** [Technologicat/mcpyrate#15]
Nested quasiquotes and staging are related. Google for `staging compilation` to find food for thought.

- Should we evaluate unquoted snippets only at `quotelevel = 0`?
  - To protect the macro bindings for the unquote operators, we could hygienify them when the outermost quote is processed.
- Multi-stage programs can typically compile and run quoted code dynamically (bringing it into the current stage). We can already do something like that, but it's not obvious how to do it. Should we provide a convenience function?

Here's a minimal example with single-phase compilation (as of 3.0.0, see the module `mcpyrate.test.test_quotes`):

```python
__all__ = ["run"]

import ast
from mcpyrate.astfixers import fix_locations
from mcpyrate.core import global_postprocess

def _fake_import(tree):
    """`tree`: list of statement AST nodes -> code object"""
    module = ast.Module(body=tree)
    reference_node = ast.Constant(value=None, lineno=1, col_offset=1)
    fix_locations(tree, reference_node, mode="reference")
    tree = global_postprocess(tree)
    code = compile(module, filename="<string>", mode="exec", dont_inherit=True)
    return code

def run(tree, globals, locals=None):
    """`tree`: list of statement AST nodes. Compile and run it with given dicts for `globals` and `locals`."""
    code = _fake_import(tree)
    if locals is not None:
        exec(code, globals, locals)
    else:
        exec(code, globals)

# ----------------------------------------

from mcpyrate.quotes import macros, q
def demo():
    with q as quoted:
        x = 42
    namespace = {}
    run(quoted, namespace)
    assert namespace["x"] == 42

if __name__ == "__main__":
    demo()
```

**Comment by Technologicat** [comment]
To facilitate a full-featured `_fake_import`, we should probably split off the part that calls the various expanders in the `source_to_xcode` function in  `mcpyrate.importer`.

**Comment by Technologicat** [comment]
As of f9b3f17, it is possible to conveniently run macro-enabled code snippets. The compiler has been refactored. See [`mcpyrate.compiler.run`](https://github.com/Technologicat/mcpyrate/blob/master/mcpyrate/compiler.py) and its friends. Usage example is in [`mcpyrate.test.test_quotes`](https://github.com/Technologicat/mcpyrate/blob/master/mcpyrate/test/test_quotes.py) for now. This feature will be released in 3.1.0.

So what remains in this issue is to think about nested quasiquotes.

**Comment by Technologicat** [comment]
Nested quasiquotes resolved in 7c08ebe.

This is the current design taken to its logical extreme. A better solution may need a break from tradition, in that maybe something being a macro should be a property of its definition site, not of its use site. Also, maybe the quasiquote operators should be considered core functionality, and not be renameable (like regular macros are).

**Comment by Technologicat** [comment]
This was released in 3.1.0.
